### PR TITLE
Building trust 0031

### PR DESCRIPTION
**Pull request by Oskar-van-Deventer**
Signed-off-by: Oskar van Deventer oskar.vandeventer@tno.nl

The Hipe on the Connection Protocol is rather parsimonious on the trust-building step. This pull request proposes to upgrade that step to a formal step. It also provides some examples of risks associated with the connection, and possible exchanges of proofs to reduce these risks.

### CODE REVIEW

File: text/0031-connection-protocol/README.md [hunk 1]
```diff
@@ -394,9 +394,28 @@ If no message is needed immediately, a trust ping can be used to allow both part
...
-#### Next Steps
+## 4. Trust Building
 
-The connection between the _inviter_ and the _invitee_ is now established. This connection has no trust associated with it. The next step should be the exchange of proofs to build trust sufficient for the purpose of the relationship.
+[4-trust-building]: #4-trust-building
+
+The connection between the *inviter* and the *invitee* is now established. This connection has no trust associated with it. The next step should be the exchange of proofs to build trust sufficient for the purpose of the relationship. The following are examples of this.
+
+#### Man-in-the-middle
+Risk: The combination of DID methods and channel(s) used in the set-up results in a too high residual risk of a man-in-the-middle. That is, the pair of DDO may be different between the two sides.
```
Comment (dhh1128) [inline text/0031-connection-protocol/README.md]: The term "DDO" is an old one that is no longer used by the community. "DID Doc" is preferred.

File: text/0031-connection-protocol/README.md [hunk 2]
```diff
@@ -394,9 +394,28 @@ If no message is needed immediately, a trust ping can be used to allow both part
...
+Mitigation: The two sides together verify the consistency of the pair of DDO pair between the two sides via an independent channel. For instance, the invitee calculates the sum of the hashes of the two DDOs, and conveys this via an independent channel to the inviter, who verifies it. Subsequently, the inviter calculates the product of the two hashes, and conveys this via an independent channel to the invitee, who verifies it. If the independent channel is a human-to-human or human-to-bot channel, then likely the first few characters of the two hashes suffice.
```
Comment (dhh1128) [inline text/0031-connection-protocol/README.md]: This is a good step but is unnecessary if the DID Doc is stored on a blockchain or other unambiguous source of truth. We should note that this is helpful for the non-blockchain DID methods.

File: text/0031-connection-protocol/README.md [hunk 3]
```diff
@@ -394,9 +394,28 @@ If no message is needed immediately, a trust ping can be used to allow both part
...
+
+#### Integrity of the hardware
+Risk: Either side may have requirements about the other side’s hardware, e.g. that a secure hardware module is used and that specific keys cannot leave the hardware module or have been added to it.
+Mitigation: Either side provides the required verifiable credentials about its secure hardware and the keys that it contains.
+
+#### Non uniqueness of DID
+Risk: An identify thief has gotten access to someone else’s DID/DDO, and tries to setup a new DID pair and to use verifiable credentials associated to the DID.
+Mitigation: Whenever an inviter or invitee unexpectedly receives a DID that is already present in an existing relationship, it should ignore the attempt and contact the other side via a sufficiently trusted channel. The other side can then take the appropriate measures, like updating the DDO and/or getting verifiable credential associated with the compromised materials revoked.
+Note: If the used DID method does not assume global uniqueness, then the new DID pair can be established as belonging to a new relationship, without any further assumptions.
```
Comment (dhh1128) [inline text/0031-connection-protocol/README.md]: I don't think this "does not assume global uniqueness" is ever the case.

File: text/0031-connection-protocol/README.md [hunk 4]
```diff
@@ -394,9 +394,28 @@ If no message is needed immediately, a trust ping can be used to allow both part
...
+
+#### Man-in-the-middle
+Risk: The combination of DID methods and channel(s) used in the set-up results in a too high residual risk of a man-in-the-middle. That is, the pair of DDO may be different between the two sides.
+Mitigation: The two sides together verify the consistency of the pair of DDO pair between the two sides via an independent channel. For instance, the invitee calculates the sum of the hashes of the two DDOs, and conveys this via an independent channel to the inviter, who verifies it. Subsequently, the inviter calculates the product of the two hashes, and conveys this via an independent channel to the invitee, who verifies it. If the independent channel is a human-to-human or human-to-bot channel, then likely the first few characters of the two hashes suffice.
+
+#### Integrity of the DDO
+Risk: Either side may have included keys in its DDO this it does not control, whereas the purpose of the relationship requires this. Ditto for service endpoints.
```
Comment (dhh1128) [inline text/0031-connection-protocol/README.md]: Can you say more about this? It is a deliberate feature of DID Docs that they can include keys other than the ones the identity owner can easily exercise (e.g., keys written on a piece of paper and locked in a vault).

File: text/0031-connection-protocol/README.md (hunk 1 above)
Comment (Oskar-van-Deventer) [reply text/0031-connection-protocol/README.md]: I shall update the Pull Request as you suggest. I am still a Github novice, so I may need to ask for your help in this.

File: text/0031-connection-protocol/README.md (hunk 2 above)
Comment (Oskar-van-Deventer) [reply text/0031-connection-protocol/README.md]: I agree with your addition. I'll provide a line like "... unnecessary if the DID Doc is stored on an unambiguous source of truth (e.g. a well governed public blockchain), ...". 

Note that "blockchain" by itself is not synonymous with "unambiguous source of truth". Any blockchain can be forked; new records may not be fully synchronised over the network; and not all blockchains are well governed.

File: text/0031-connection-protocol/README.md (hunk 3 above)
Comment (Oskar-van-Deventer) [reply text/0031-connection-protocol/README.md]: I can remove this note for now.

File: text/0031-connection-protocol/README.md (hunk 4 above)
Comment (Oskar-van-Deventer) [reply text/0031-connection-protocol/README.md]: We are in the set-up phase. One side may suspect that the other side has been compromised or that a MITM is active, e.g. based on the type of DID and/or set-up channels used. For instance, if the other side uses the method did:stupid, a future DID method. The existence of a "proof-of-control" method suggests to me that there are cases where one side may need such proof.

**Pull request merged**